src/components/editor-page/editor-pane/hooks/use-on-image-upload-from-renderer.ts

```diff
@@ -0,0 +1,120 @@
+import type { ImageUploadMessage } from '../../../render-page/window-post-message-communicator/rendering-message'
+import { CommunicationMessageType } from '../../../render-page/window-post-message-communicator/rendering-message'
+import { useCallback } from 'react'
+import { store } from '../../../../redux'
+import { handleUpload } from '../upload-handler'
+import type { Editor, Position } from 'codemirror'
+import { Logger } from '../../../../utils/logger'
+import { findRegexMatchInLine } from '../find-regex-match-in-line'
+import Optional from 'optional-js'
+
+const log = new Logger('useOnImageUpload')
+const imageWithPlaceholderLinkRegex = /!\[([^\]]*)]\(https:\/\/([^)]*)\)/g
+
+/**
+ * Receives {@link CommunicationMessageType.IMAGE_UPLOAD image upload events} via iframe communication and processes the attached uploads.
+ *
+ * @param editor The {@link Editor codemirror editor} that should be used to change the markdown code
+ */
+export const useOnImageUploadFromRenderer = (editor: Editor | undefined): void => {
+  useEditorReceiveHandler(
+    CommunicationMessageType.IMAGE_UPLOAD,
+    useCallback(
+      (values: ImageUploadMessage) => {
+        const { dataUri, fileName, lineIndex, placeholderIndexInLine } = values
+        if (!editor) {
+          return
+        }
+        if (!dataUri.startsWith('data:image/')) {
+          log.error('Received uri is no data uri and image!')
+          return
+        }
+
+        fetch(dataUri)
+          .then((result) => result.blob())
+          .then((blob) => {
+            const file = new File([blob], fileName, { type: blob.type })
+            const { cursorFrom, cursorTo, description, additionalText } = Optional.ofNullable(lineIndex)
+              .map((actualLineIndex) => findPlaceholderInMarkdownContent(actualLineIndex, placeholderIndexInLine))
+              .orElseGet(() => calculateInsertAtCurrentCursorPosition(editor))
+            handleUpload(file, editor, cursorFrom, cursorTo, description, additionalText)
+          })
+          .catch((error) => log.error(error))
+      },
+      [editor]
+    )
+  )
+}
+
+export interface ExtractResult {
+  cursorFrom: Position
+  cursorTo: Position
+  description?: string
+  additionalText?: string
+}
+
+/**
+ * Calculates the start and end cursor position of the right image placeholder in the current markdown content.
+ *
+ * @param lineIndex The index of the line to change in the current markdown content.
+ * @param replacementIndexInLine If multiple image placeholders are present in the target line then this number describes the index of the wanted placeholder.
+ * @return the calculated start and end position or undefined if no position could be determined
+ */
+const findPlaceholderInMarkdownContent = (lineIndex: number, replacementIndexInLine = 0): ExtractResult | undefined => {
+  const currentMarkdownContentLines = store.getState().noteDetails.markdownContent.split('\n')
```

Comment:
Just noting (no change request): We should probably really cache the content split by lines. This is third or fourth occurrence of splitting the content by linebreaks.

Reply:
That's right.

Reply:
See https://github.com/hedgedoc/react-client/pull/1681